src/shims/unix/fs.rs

```diff
@@ -61,7 +57,22 @@ impl FileDescription for FileHandle {
         _ecx: &mut MiriInterpCx<'tcx>,
     ) -> InterpResult<'tcx, io::Result<usize>> {
         assert!(communicate_allowed, "isolation should have prevented even opening a file");
-        Ok(self.file.read_at(bytes, offset))
+        #[cfg(unix)]
+        {
+            use std::os::unix::fs::FileExt;
+            Ok(self.file.read_at(bytes, offset))
+        }
+        // Emulate pread using seek + read + seek to restore cursor position.
+        // Correctness of this emulation relies on sequential nature of Miri execution.
```

Comment:
What exactly is the problem with always using the portable emulation?

Reply:
Why use a subpar emulation when we have access to the original?

Reply:
One shared codepath for everything means better test coverage. The possibility of subtle breakage is an argument *for* a single implementation.

In which sense is the Unix user experience pessimized? The number of syscalls is really irrelevant, Miri is so much slower than native execution that performance is not a valid argument.

Reply:
>  I do not care about Windows targets at all

This is about windows *hosts*. Windows targets don't have any FS emulation, so they must currently use Linux targets when using Miri with programs that access the file system. That makes it even more relevant to have a maximally portable implementation in Miri.

Reply:
Personally, I would prefer to not have this emulation in the first place. The `is_symlink` emulation is not 100% equivalent to the `O_NOFOLLOW` flag, e.g. `is_symlink` may return `true` in case of an IO error, while `open` will succeed. In my opinion, it's better to to deal with reliable and straightforward shims instead of potentially problematic emulations.

If you insist on using the emulation for all hosts, I will do it.

>In which sense is the Unix user experience pessimized?

The requirement of having a Windows emulation means that it's not trivial to add UNIX APIs which do not have direct analogs in Windows. It means that it requires much more effort to implement a shim, or that it will not be implemented at all, thus limiting Miri usefulness for UNIX users.

Arguably, the burden of implementing emulation should be on users who need it. Lack of emulation does not mean that Miri is uselles, only that you need a UNIX host (or WSL) to run Miri tests for trickier programs.

Reply:
It seems you partially mixed this up with the other PR? There's no symlinks here...

> The requirement of having a Windows emulation means that it's not trivial to add UNIX APIs which do not have direct analogs in Windows. It means that it requires much more effort to implement a shim, or that it will not be implemented at all, thus limiting Miri usefulness for UNIX users.

Cross-target-support is a pretty nifty feature of Miri and increases its usefulness. So there's a tradeoff here. If something is impossible to implement or requires unreasonable effort, we have [accepted](https://github.com/rust-lang/miri/blob/c9c887b7170ef3fdc608ae7f876dac7782e92cf0/src/shims/unix/fs.rs#L327-L332) `cfg(unix)` code in the past and will do so again. But if it is relatively simple then we take the portable route.

Reply:
>It seems you partially mixed this up with the other PR? There's no symlinks here...

Yes, my mistake. But the same applies to the seek-based emulation. For example, because of race conditions it may fail where `pread`/`pwrite` would've succeeded or return error codes may be different. There is less probability for something going wrong with a straightforward shim. The testing argument means that you trade experience of most users for a probability of improving experience for a minority.

Do you insist on using the emulation for all hosts (in this and in the `O_NOFOLLOW` PR)?

>Cross-target-support is a pretty nifty feature of Miri and increases its usefulness.

I do not argue against its usefulness. My point is that such emulations should be done by users who need them. I find it unlikely that Windows developers will test programs which rely on `pread`/`pwrite` or `O_NOFOLLOW`.

Reply:
> because of race conditions

Which race conditions? Miri is sequential.

> Do you insist on using the emulation for all hosts (in this and in the O_NOFOLLOW PR)?

This is a case-by-case thing. Currently I am trying to understand the tradeoff for this PR.

Reply:
> because of race conditions

Are you trying to talk about a race on the file descriptor's state?

Reply:
>Which race conditions? Miri is sequential.

With outside environment. For example, if we test a program which works in tandem with other program.

In general, I do not want to think about potential ways for the emulation to break or have an observably different behavior. But it's more important for me to have `pread`/`pwrite` support in Miri, so if using the emulation for all hosts will make acceptance of this PR easier and quicker, I am fine with changing the code accordingly. I will create a commit with such changes, if you'll decide that we do not need emulation on UNIX hosts after all, we can revert it.

Reply:
I am trying to understand which downsides you are seeing. The outside environment cannot observe or change our file descriptor state (Miri doesn't do FD sharing with other processes) so I fail to see the possible problems there.

But if you prefer we can also skip this and just go with seek-based emulation on all hosts.